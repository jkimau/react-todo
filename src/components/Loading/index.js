import React from 'react';
import styled from 'styled-components';

const LoadingDiv = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  margin-top: -9px;
  margin-left: -35px;
`;

const Loading = () => (
  <LoadingDiv>Loading...</LoadingDiv>
);

export default Loading;
