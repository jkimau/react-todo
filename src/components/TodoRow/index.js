import React, { Component } from 'react';
import ReactCSSTransitionGroup from 'react-addons-css-transition-group';
import styled from 'styled-components';

import { mainBG, todoRowBorder } from 'global/colors';

const padding = `15px`;

const TodoRowWrapper = styled.div`
  position: relative;
  background: white;
  border-bottom: 1px solid ${todoRowBorder};
  color: ${props => props.completed ? '#ccc' : 'inherit'};

  &:last-child {
    border-bottom: none;
  }

  &:hover {
    background: ${mainBG};
  }
`;

const Header = styled.div`
  padding: ${padding}
`;

const Body = styled.div`
  padding: ${padding};
`;

const OpenClose = styled.span`
  float: right;
  cursor: pointer;
`;

const TodoRowMenu = styled.ul`
  position: absolute;
  top: 45px;
  right: -100px;
  z-index: 100;
  margin: 0;
  width: 160px;
  border-radius: 4px;
  list-style: none;
  padding-left: 0;
  background: white;
  color: #666;
  overflow: hidden;

  &.slide-enter {
    opacity: 0;
    height: 0px;
    border: none;
    transition: all 0.05s ease;
  }

  &.slide-leave {
    opacity: 1;
    height: ${props => props.totalHeight ? `${props.totalHeight}px` : '100px'};
    border: 1px solid #ddd;
    transition: all 0.05s ease;
  }

  &:focus {
    outline: none;
  }

  li {
    height: ${props => props.listHeight ? `${props.listHeight}px` : 'auto'};
    padding: 6px;
    box-sizing: border-box;
    text-align: center;
    font-size: 14px;
    cursor: pointer;

    &:hover {
      background: #2FC2EF;
      color: white;
    }
  }
`;

export default class TodoRow extends Component {
  constructor(props) {
    super(props);

    this.toggleMenu = this.toggleMenu.bind(this);
  }

  toggleMenu(e) {
    const { todo: { id }, toggleRowMenuHandler } = this.props;
    toggleRowMenuHandler(id);
  }

  render() {
    const {
      todo: {
        title,
        date,
        details,
        completed,
        menuOpen
      }
    } = this.props;
    return (
      <TodoRowWrapper completed={completed}>
        <Header>
          <span>{title}</span>
          <span>{date}</span>
          <OpenClose onClick={this.toggleMenu} className="todo-menu-trigger">
            {menuOpen ? 'close' : 'open'}
          </OpenClose>
          <ReactCSSTransitionGroup
            transitionName="slide"
            transitionEnterTimeout={0}
            transitionLeaveTimeout={0}>
            <TodoRowMenu
              className={menuOpen ? 'slide-leave' : 'slide-enter'}
              totalHeight={30 * 2}
              listHeight={30}>
              <li>Mark complete</li>
              <li>Delete</li>
            </TodoRowMenu>
          </ReactCSSTransitionGroup>
        </Header>
        <Body>{details}</Body>
      </TodoRowWrapper>
    );
  }
}
