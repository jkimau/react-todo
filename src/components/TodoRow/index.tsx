import React, { Component } from 'react';
import styled from 'styled-components';

import { mainBG, todoRowBorder } from 'global/colors';
import { TodoPropTypes } from 'propTypes';

const padding = `15px`;

const TodoRowWrapper = styled.div`
  position: relative;
  background: white;
  border-bottom: 1px solid ${todoRowBorder};
  color: ${props => props.completed ? '#ccc' : 'inherit'};

  &:last-child {
    border-bottom: none;
  }

  &:hover {
    background: ${mainBG};
  }
`;

const Header = styled.div`
  padding: ${padding}
`;

const Body = styled.div`
  padding: ${padding};
`;

const OpenClose = styled.span`
  float: right;
  cursor: pointer;
`;

const TodoRowMenu = styled.ul`
  position: absolute;
  top: 45px;
  right: -100px;
  z-index: 100;
  margin: 0;
  width: 160px;
  border-radius: 4px;
  list-style: none;
  padding-left: 0;
  background: white;
  color: #666;
  overflow: hidden;

  &.slide-enter {
    opacity: 0;
    height: 0px;
    border: none;
    transition: all 0.05s ease;
  }

  &.slide-leave {
    opacity: 1;
    height: ${props => props.totalHeight ? `${props.totalHeight}px` : '100px'};
    border: 1px solid #ddd;
    transition: all 0.05s ease;
  }

  &:focus {
    outline: none;
  }

  li {
    height: ${props => props.listHeight ? `${props.listHeight}px` : 'auto'};
    padding: 6px;
    box-sizing: border-box;
    text-align: center;
    font-size: 14px;
    cursor: pointer;

    &:hover {
      background: #2FC2EF;
      color: white;
    }
  }
`;

interface TodoRowPropTypes {
  key: number,
  todo: TodoPropTypes,
  toggleTodoMenu(id: number): void
};

export default class TodoRow extends Component<TodoRowPropTypes, null> {
  constructor(props) {
    super(props);

    this.toggleTodoMenu = this.toggleTodoMenu.bind(this);
    this.stopPropagateEvent = this.stopPropagateEvent.bind(this);
  }

  toggleTodoMenu() {
    this.props.toggleTodoMenu(this.props.todo.id);
  }

  stopPropagateEvent(e) {
    e.stopPropagation();
  }

  render() {
    const {
      todo: {
        title,
        date,
        details,
        completed,
        menuOpen
      }
    } = this.props;

    return (
      <TodoRowWrapper completed={completed}>
        <Header>
          <span className="todo-row-header">{title}</span>
          <span className="todo-row-date">{date}</span>
          <OpenClose onClick={this.toggleTodoMenu} className="todo-menu-trigger">
            {menuOpen ? 'close' : 'open'}
          </OpenClose>
          <TodoRowMenu
            onClick={this.stopPropagateEvent}
            className={`todo-row-menu ${menuOpen ? 'slide-leave' : 'slide-enter'}`}
            totalHeight={30 * 2}
            listHeight={30}>
            <li>Mark complete</li>
            <li>Delete</li>
          </TodoRowMenu>
        </Header>
        <Body className="todo-row-details">{details}</Body>
      </TodoRowWrapper>
    );
  }
}
