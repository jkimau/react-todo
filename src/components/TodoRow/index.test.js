import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, render, mount } from 'enzyme';

import TodoRow from '../TodoRow';
import Loading from 'components/Loading';

jest.mock('global/colors', () => ({
  mainBG: '#aaa',
  todoRowBorder: '#ccc'
}));

describe('<TodoRow/>', () => {
  const defaultProps = {
    todo: {
      title: 'test title',
      date: '2017-06-01T00:00:00Z',
      details: 'test details',
      id: 1,
      menuOpen: false,
      completed: false
    },
    toggleTodoMenu: jest.fn()
  }

  /**
  * Snapshot
  */
  describe(' -- snapshot', () => {
    test('snapshot with menuOpen true', () => {
      const tree = renderer.create(
        <TodoRow {...defaultProps}/>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    test('snapshot with menuOpen close', () => {
      const newProps = Object.assign({}, defaultProps);
      newProps.todo = Object.assign({}, defaultProps.todo);
      newProps.todo.menuOpen = true;

      const tree = renderer.create(
        <TodoRow {...newProps}/>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  /**
  * Shallow render
  */
  describe(' -- shallow render React', () => {
    const compileComponent = (props) => {
      return shallow(<TodoRow {...props}/>);
    }

    test('should render texts properly', () => {
      const wrapper = compileComponent(defaultProps);

      expect(wrapper.find('.todo-row-header').text()).toEqual('test title');
      expect(wrapper.find('.todo-row-date').text()).toEqual('2017-06-01T00:00:00Z');

      // above 2 are normal html element whereas .todo-row-details is a styled-component
      expect(wrapper.render().find('.todo-row-details').text()).toEqual('test details');
    });

    test('should todo row menu has className(1)', () => {
      const wrapper = compileComponent(defaultProps);

      expect(wrapper.find('.todo-row-menu').hasClass('slide-enter')).toBe(true);
    });

    test('should todo row menu has className(2)', () => {
      const newProps = Object.assign({}, defaultProps);
      newProps.todo = Object.assign({}, defaultProps.todo);
      newProps.todo.menuOpen = true;
      const wrapper = compileComponent(newProps);

      expect(wrapper.render().find('.todo-row-menu').hasClass('slide-leave')).toBe(true);
    });

    test('should call toggleTodoMenu()', () => {
      const wrapper = compileComponent(defaultProps);
      wrapper.find('.todo-menu-trigger').simulate('click');

      expect(wrapper.instance().props.toggleTodoMenu).toHaveBeenCalled();
    });

    test('should call stopPropagateEvent()', () => {
      TodoRow.prototype.stopPropagateEvent = jest.fn();
      const wrapper = compileComponent(defaultProps);
      wrapper.find('.todo-row-menu').simulate('click');

      expect(TodoRow.prototype.stopPropagateEvent).toHaveBeenCalled();
    });
  });
});
