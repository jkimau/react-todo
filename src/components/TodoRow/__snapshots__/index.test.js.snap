exports[`<TodoRow/>  -- snapshot snapshot with menuOpen close 1`] = `
<div
  className="ihdFiv">
  <div
    className="Eszii">
    <span
      className="todo-row-header">
      test title
    </span>
    <span
      className="todo-row-date">
      2017-06-01T00:00:00Z
    </span>
    <span
      className="todo-menu-trigger bPMTRm"
      onClick={[Function]}>
      close
    </span>
    <ul
      className="todo-row-menu slide-leave gHZgXS"
      onClick={[Function]}>
      <li>
        Mark complete
      </li>
      <li>
        Delete
      </li>
    </ul>
  </div>
  <div
    className="todo-row-details hAJdRI">
    test details
  </div>
</div>
`;

exports[`<TodoRow/>  -- snapshot snapshot with menuOpen true 1`] = `
<div
  className="ihdFiv">
  <div
    className="Eszii">
    <span
      className="todo-row-header">
      test title
    </span>
    <span
      className="todo-row-date">
      2017-06-01T00:00:00Z
    </span>
    <span
      className="todo-menu-trigger bPMTRm"
      onClick={[Function]}>
      open
    </span>
    <ul
      className="todo-row-menu slide-enter gHZgXS"
      onClick={[Function]}>
      <li>
        Mark complete
      </li>
      <li>
        Delete
      </li>
    </ul>
  </div>
  <div
    className="todo-row-details hAJdRI">
    test details
  </div>
</div>
`;
